client.go

```diff
@@ -2294,6 +2294,105 @@ func (c *ToznySDKV3) UnshareSecretFromUsername(ctx context.Context, options Unsh
 	return nil
 }
 
+type UnshareBeforeDeleteOptions struct {
+	SecretID       uuid.UUID
+	CallerClientID uuid.UUID
+	Type           string
+}
+
+// UnshareSecretBeforeDelete unshares the secret with SecretID from every group it's shared with
+// and deletes the group if it contains no other secrets.
+func (c *ToznySDKV3) UnshareSecretBeforeDelete(ctx context.Context, options UnshareBeforeDeleteOptions) ([]error, error) {
+	listRequest := storageClient.ListGroupsRequest{
+		ClientID: options.CallerClientID,
+	}
+	listGroupResponse, err := c.StorageClient.ListGroups(ctx, listRequest)
+	if err != nil {
+		return nil, err
+	}
+	secretID := options.SecretID
+	var processingErrors []error
+	// Check each group the calling client belongs to for the secret
+	for _, group := range listGroupResponse.Groups {
+		if !ValidToznySecretNamespace(group.Name) {
+			continue
+		}
+		listRequest := storageClient.ListGroupRecordsRequest{
+			GroupID:   group.GroupID,
+			WriterIDs: []string{options.CallerClientID.String()},
+		}
+		// Get all the records shared with the group
+		listGroupRecords, err := c.StorageClient.GetSharedWithGroup(ctx, listRequest)
+		if err != nil {
+			msg := fmt.Errorf("UnshareSecretBeforeDelete: could not access group %s. Err: %+v", group.GroupID, err)
+			processingErrors = append(processingErrors, msg)
+			continue
+		}
+		numberRecordsInGroup := len(listGroupRecords.ResultList)
+		// unshare the record from the group
+		recordRemoveShareRequest := storageClient.RemoveRecordSharedWithGroupRequest{
+			GroupID:    group.GroupID,
+			RecordType: options.Type,
+		}
+		err = c.StorageClient.RemoveSharedRecordWithGroup(ctx, recordRemoveShareRequest)
+		if err != nil {
+			msg := fmt.Errorf("UnshareSecretBeforeDelete: failed to remove secret %s from group %s. Err: %+v", secretID, group.GroupID, err)
+			processingErrors = append(processingErrors, msg)
```

Comment:
should we continue here or will the calls below be happy if we ran into an error removing the secret from the group?

Reply:
Good point. I don't think it would break anything without the continue -- since the DeleteGroup will also fail -- but it does probably make more sense to continue here. I'll add that.